[Android] Decouple runtime initialization and entry point execution for Android sample

matouskozak opened the pull request:
# Problem Description
On Android, the startup is measured by scanning the logcat output for the following message:

```
Displayed <name> for user 0: <time>
```

In the case of dotnet/runtime sample app it would be e.g.: 
```
Displayed net.dot.HelloAndroid/net.dot.MainActivity for user 0: +505ms
```
More information regarding startup measurements can be found at https://developer.android.com/topic/performance/vitals/launch-time#retrieve-TTID.

In the previous design, the `onCreate` method would finish before the runtime would get initialized (`MonoRunner.initialize`). This means that when we're reading the `Displayed <name> for user 0: <time>`, only the Java code had run, rendering all startup measurements useless as we couldn't measure any impact from our runtime.

# Proposed Solution
By decoupling the initialization of runtime and execution of entry point, we can run the runtime initialization synchronously, forcing the `onCreate` callback to report the `Displayed <name> for user 0: <time>` after the runtime is fully initialized. 

The execution of the entry point is still performed asynchronously to prevent the freeze of main thread for strenuous executables.

Note, this is solution doesn't replicate real-world scenarios, but it allows us to measure the startup time accurately with only the dotnet/runtime code. 

---

## Future Work
- unify code style between the `monodroid.c` and `monodroid-coreclr.c` files
- extract shared functionality (https://github.com/dotnet/runtime/issues/111923)
- figure out appropriate hosting of CoreCLR (and Mono)

akoeplinger commented:
If I remember correctly this was intentional since `MonoRunner.initialize` doesn't just initialize the runtime, it essentially runs the full program until it exits so doing that in onCreate would freeze the UI if the program takes too long.

matouskozak commented:
> If I remember correctly this was intentional since `MonoRunner.initialize` doesn't just initialize the runtime, it essentially runs the full managed program until it exits so doing that in onCreate would freeze the UI if the program takes too long.

Indeed, that is what I observed as well. The startup time reported would be more TTFD than TTID. We could probably split https://github.com/dotnet/runtime/blob/c74440f8291edd35843f3039754b887afe61766e/src/tasks/AndroidAppBuilder/Templates/monodroid.c#L308-L318 to a separate function call and call just this one asynchronously.

matouskozak commented:
/azp run runtime-androidemulator

matouskozak committed:
split init runtime and execution of entry point

matouskozak commented:
/azp run runtime-androidemulator

matouskozak commented:
/azp run runtime-android

akoeplinger commented on src/tasks/AndroidAppBuilder/Templates/monodroid.c:
```diff
@@ -304,7 +307,11 @@ mono_droid_runtime_init (const char* executable, int managed_argc, char* managed
 
+static int 
+mono_droid_execute_assembly (const char *executable, int managed_argc, char* managed_argv[])
+{    
```
whitespace:
```suggestion
{
```

akoeplinger commented on src/tasks/AndroidAppBuilder/Templates/monodroid.c:
```diff
@@ -304,7 +307,11 @@ mono_droid_runtime_init (const char* executable, int managed_argc, char* managed
+static int 
+mono_droid_execute_assembly (const char *executable, int managed_argc, char* managed_argv[])
+{    
     MonoDomain *domain = mono_jit_init_version ("dotnet.android", "mobile");
```
I think we can move mono_jit_init_version and mono_droid_load_assembly into the runtime init too, those don't execute user code and I'd count them as startup.

matouskozak replied:
I thought keeping it with `mono_droid_execute_assembly` because the `domain` and `assembly` variables are needed for `mono_jit_exec` call. If we would keep them separate, I would have to marshal them to java and then back to c (or save them somewhere and load them for the call). Or do you have a better suggestion how this could be handled?

Also `mono_droid_load_assembly` seemed like it is more related to execution of user code than runtime init.

matouskozak replied:
Based on the CoreCLR changes, I probably should move them to runtime init as suggested to be aligned.

matouskozak replied:
Moved to `mono_droid_runtime_init`

akoeplinger commented on src/tasks/AndroidAppBuilder/Templates/monodroid.c:
```diff
@@ -33,8 +33,11 @@
+Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_dir, jstring j_cache_dir, jstring j_testresults_dir, long current_local_time);
+
 int
-Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_dir, jstring j_cache_dir, jstring j_testresults_dir, jstring j_entryPointLibName, jobjectArray j_args, long current_local_time);
```
@steveisok does this affect librarymode? I don't really see how this works

ivanpovazan replied:
Library mode uses "on library load" runtime initialization and is built as part of the lib: https://github.com/dotnet/runtime/blob/0a477a86d0fea6234ada63e660c0faa310fe121d/src/tasks/LibraryBuilder/Templates/autoinit.c#L164 

and a separate monodroid source is used: https://github.com/dotnet/runtime/blob/0a477a86d0fea6234ada63e660c0faa310fe121d/src/tasks/AndroidAppBuilder/Templates/monodroid-librarymode.c 

so if we are changing the JNI method signature, that one ^ should also be updated.

@matouskozak I added a library mode test in: https://github.com/dotnet/runtime/pull/111545 and would put that PR also as a prerequisite to this one, so we have a better test coverage.

Finally, I would treat library mode as out of scope.

vitek-karas commented:
Just curious: I understand that with this change we will now measure startup as "the time it takes for the runtime to init far enough to be able to run some code" - but it won't run "Main" yet. I'm afraid that comparing this to CoreCLR will be tricky, since the two runtimes will probably do different things in "init" versus "run main". For example, I'm pretty sure CoreCLR actually runs managed code during "init". It's possible mono delays it until "run main".

I think a much more representative measurement would be to log something from managed Main - and measure how long it takes to do that - in that case we're measuring the same behavior not relying on runtime specific implementation differences.

matouskozak commented:
> I think a much more representative measurement would be to log something from managed Main - and measure how long it takes to do that - in that case we're measuring the same behavior not relying on runtime specific implementation differences.

I would rather avoid any custom logging inside the dotnet/runtime sample. By using custom log message, we would have to frame the performance startup scripts around the custom logging. Currently the scripts work by detecting the `Displayed <name> for user 0: <time>` which is a [default standard](https://developer.android.com/topic/performance/vitals/launch-time#retrieve-TTID) working with any APK file.

> Just curious: I understand that with this change we will now measure startup as "the time it takes for the runtime to init far enough to be able to run some code" - but it won't run "Main" yet. I'm afraid that comparing this to CoreCLR will be tricky, since the two runtimes will probably do different things in "init" versus "run main".

I definitely agree with this part. Right now, I'm not 100% sure how to solve this other than synchronizing both runtime initialization and execution of main. 

>  For example, I'm pretty sure CoreCLR actually runs managed code during "init". It's possible mono delays it until "run main".

After the https://github.com/dotnet/runtime/pull/110471 gets merged I can run it locally for CoreCLR and see what numbers we get.

matouskozak commented:
One posibility to explore is using [TTFD](https://developer.android.com/topic/performance/vitals/launch-time#time-full) for the dotnet/runtime sample instead of TTID. We could call the `reportFullyDrawn()` after the completion of main execution and use this time as "startup" for dotnet/runtime Android sample.

TTID: `Displayed net.dot.HelloAndroid/net.dot.MainActivity for user 0: +433ms`
TTFD: `Fully drawn net.dot.HelloAndroid/net.dot.MainActivity: +1s397ms`

ivanpovazan commented:
> Just curious: I understand that with this change we will now measure startup as "the time it takes for the runtime to init far enough to be able to run some code" - but it won't run "Main" yet. I'm afraid that comparing this to CoreCLR will be tricky, since the two runtimes will probably do different things in "init" versus "run main". For example, I'm pretty sure CoreCLR actually runs managed code during "init". It's possible mono delays it until "run main".

The comparison won't be 100% "fair" and that is fine. We use `Hello*` samples measurements to get the general idea about performance between runtimes, while the actual *real* measurement/comparison comes with `dotnet new maui/ios/android` samples, where full target-language interop is included. Additionally, we get `Hello*` measurements much sooner as we depend only on dotnet/runtime, so their "real" value is in early regression detection. 

One example of another "unfair" measurement is for size for HelloiOS apps between NativeAOT and Mono where on Mono we can't strip symbols in the same way: and this gives us a comparison of: 1MB vs 5MB, which isn't really what we are getting in reality - MAUI apps :)

matouskozak committed:
Merge branch 'main' into synchronize-runtime-init-android

matouskozak committed:
add TTFD callback

matouskozak committed:
introduce parity to mono library mode

AaronRobinsonMSFT commented:
/cc @elinor-fung

matouskozak committed:
adapt coreclr android to separete init and exec

matouskozak committed:
adapt monodroid.c to have same signature as clr

matouskozak committed:
library mode signature updates

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -24,12 +24,21 @@ void
 
-static char *bundle_path;
-static char *executable;
+static char* g_bundle_path = NULL;
+static char* g_executable_path = NULL;
```
```suggestion
static const char* g_bundle_path = NULL;
static const char* g_executable_path = NULL;
```

matouskozak replied:
Thanks, edited for both Mono and CoreCLR.

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -198,25 +239,61 @@ Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_
 
-    bundle_path = file_dir;
-    executable = entryPointLibName;
+    g_bundle_path = (char*)malloc(sizeof(char) * (strlen(file_dir) + 1)); // +1 for '\0'
+    if (g_bundle_path == NULL)
+    {
+        LOG_ERROR("Failed to allocate memory for bundle_path");
+        return -1;
+    }
+    strncpy(g_bundle_path, file_dir, strlen(file_dir) + 1);
```
```suggestion
    size_t file_dir_len = strlen(file_dir);
    char* bundle_path_tmp = (char*)malloc(sizeof(char) * (file_dir_len + 1)); // +1 for '\0'
    if (bundle_path_tmp == NULL)
    {
        LOG_ERROR("Failed to allocate memory for bundle_path");
        return -1;
    }
    strncpy(bundle_path_tmp, file_dir, file_dir_len + 1);
    g_bundle_path = bundle_path_tmp;
```

matouskozak replied:
Thanks, edited for both Mono and CoreCLR.

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -130,9 +129,58 @@ get_tpas_from_path(const char* dir_path, const char** tpas)
 
 static int
-mono_droid_runtime_init (const char* executable, int managed_argc, char* managed_argv[], int local_date_time_offset)
+bundle_executable_path (const char* executable, const char* bundle_path, char** executable_path)
```
```suggestion
bundle_executable_path (const char* executable, const char* bundle_path, const char** executable_path)
```

matouskozak replied:
Thank you, I bundled the changes with other refactoring and addressing of feedback.

matouskozak committed:
apply review feedback
Move mono_jit_init_version and mono_droid_load_assembly to mono_droid_runtime_init
type safety around strings

matouskozak committed:
move resource cleanup to onDestroy

matouskozak commented:
/azp run runtime-android,runtime-androidemulator

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -130,9 +129,64 @@ get_tpas_from_path(const char* dir_path, const char** tpas)
+        free (g_executable_path);
+        g_executable_path = NULL;
+    }
+    if (g_coreclr_handle)
+    {
+        coreclr_shutdown (g_coreclr_handle, g_coreclr_domainId);
+        g_coreclr_handle = NULL;
+    }
```
Just an FYI. Calling `coreclr_shutdown` doesn't really make the runtime unloadable. I personally think we should avoid calling this API in general because it isn't like we can reinitialize CoreCLR or unload anything. It is kind of bookkeeping and in general leads to a false sense of "clean state".

You can absolutely leave it in, but it doesn't really do what one might think it does.

matouskozak replied:
Good point. My assumption was that this should be called after the workload is finished. If it's not needed I'll remove it for simplicity.

matouskozak replied:
Thinking about this, is there a proper way to free the resources hold by `g_coreclr_handle` or is this something we don't need to worry about?

AaronRobinsonMSFT replied:
This is the only API that releases many of those resources. Releasing them is completely reasonable. My main points with this comment were that (a) this doesn't actually shutdown the runtime in a way that makes it re-activatable or unloadable - that isn't intuitive based on its name, and (b) CoreCLR generally assumes it "owns" the process, in this case it doesn't, and this API will only be called on the way to process termination.

Based on what I'm seeing, this path is only called during the process termination sequence which makes it completely reasonable to keep. My suggestion would simply be to add a comment such as:

> Clean up some coreclr resources. This doesn't make coreclr unloadable.

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -130,9 +129,64 @@ get_tpas_from_path(const char* dir_path, const char** tpas)
+    unsigned int rv;
+    LOG_INFO ("Calling coreclr_execute_assembly");
+    coreclr_execute_assembly (coreclr_handle, coreclr_domainId, managed_argc, managed_argv, executable_path, &rv);
+    LOG_INFO ("Exit code: %d.", rv);
```
```suggestion
    LOG_INFO ("Exit code: %u.", rv);
```

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -130,9 +129,64 @@ get_tpas_from_path(const char* dir_path, const char** tpas)
+static int
+mono_droid_runtime_init (const char* bundle_path, const char* executable, int local_date_time_offset)
+{
+    LOG_INFO ("mono_droid_runtime_init (CoreCLR) called with executable: %s", executable);
```
Is the `bundle_path` important here?

matouskozak replied:
Not important, we can use the global `g_bundle_path` instead.

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -153,42 +207,35 @@ mono_droid_runtime_init (const char* executable, int managed_argc, char* managed
+void
+Java_net_dot_MonoRunner_setEnv (JNIEnv* env, jobject thiz, jstring j_key, jstring j_value)
+{
+    LOG_INFO ("Java_net_dot_MonoRunner_setEnv:");
```
I would assert that coreclr hasn't been initialized yet. In fact, it might make sense to return a non-zero value if someone calls this and coreclr has been initialized. The `setenv` is a tricky API and rather untrustworthy from a threading perspective.

matouskozak replied:
Why is it important that the runtime hasn't been initialized yet for setting env variables?

AaronRobinsonMSFT replied:
My assumption here is that this is primarily for setting environment variables for use during coreclr initialization, that could be wrong. The issue with that is there are some environment variables that aren't obvious when they are read/consumed and it can cause confusion if all `DOTNET_` related variables are set a priori.

On a general note for `setenv`, it isn't thread safe. In coreclr we have a big lock around it for that very reason. However, we can't control other callers of that API and one can get into weird situation when reading the environment block and other threads are setting it. This could have been addressed by the Android system, but as far as I'm aware it is still a common issue with the API not being thread safe.

AaronRobinsonMSFT commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -198,25 +245,57 @@ Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_
+    g_bundle_path = bundle_path_tmp;
 
-    setenv ("HOME", bundle_path, true);
+    setenv ("HOME", g_bundle_path, true);
     setenv ("TMPDIR", cache_dir, true);
     setenv ("TEST_RESULTS_DIR", testresults_dir, true);
```
Should we just set these in Java and avoid doing this here?

matouskozak commented on src/tasks/AndroidAppBuilder/Templates/MainActivity.java:
```diff
@@ -41,11 +41,13 @@ protected void onCreate(Bundle savedInstanceState)
         }
 
         final Activity ctx = this;
+        MonoRunner.initializeRuntime(entryPointLibName, ctx);
```
Based on my understanding, both `initializeRuntime` and the handler below will run on the same thread (cc: @jonathanpeppers). The motivation behind splitting and using the post handler is that the `onCreate` from MainActivity can finish and report "Displayed..." to the logcat which is captured for TTID measurement. Currently, the TTID should include all the Java parts + the `initializeRuntime`. The post handler below will execute with a delay (after `onCreate` finished) and run the executable.

jonathanpeppers replied:
I wondered if you should remove the `new Handler(Looper.getMainLooper()).postDelayed(new Runnable()` call completely. .NET for Android apps do not call `post()` to run user code on startup.

This also would make TTFD less important, and maybe you wouldn't even need to record that number. TTFD will include time Android spends doing things that we have little control over.

The reason you have for using `postDelayed()` is:

> The execution of the entry point is still performed asynchronously to prevent the freeze of main thread for strenuous executables.

We are doing this in the product in customers' apps, so seems reasonable to do it in this sample as well.

Removing `postDelayed()` could be done in another PR, though.

matouskozak replied:
Thanks @jonathanpeppers for the comment and insights.

Indeed, .NET Android follows a different approach to how the runtime is initialized and code ran but I wasn't sure if the effort to rewrite the dotnet/runtime sample is worth for our use case. Splitting the initialization and execution of managed code seemed like a good compromise to get some meaningful information from TTID and TTFD. However, we could definitely create a proposal for the re-work if that's wanted.

AaronRobinsonMSFT reviewed (approved):
LGTM.

ivanpovazan reviewed (commented):
LGTM, nice work! :)

matouskozak commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -24,12 +24,21 @@ void
+Java_net_dot_MonoRunner_execEntryPoint (JNIEnv* env, jobject thiz, jstring j_entryPointLibName, jobjectArray j_args);
+
+void
+Java_net_dot_MonoRunner_freeNativeResources (JNIEnv* env, jobject thiz);
```
Yes, I already suggested it in the issue description as future work. Do you want me to open an issue for this after this PR gets merged?

ivanpovazan replied:
https://github.com/dotnet/runtime/issues/111923 added it to the TODO list there.

ivanpovazan commented on src/tasks/AndroidAppBuilder/Templates/monodroid-coreclr.c:
```diff
@@ -54,21 +63,11 @@ strncpy_str (JNIEnv *env, char *buff, jstring str, int nbuff)
     jboolean isCopy = 0;
     const char *copy_buff = (*env)->GetStringUTFChars (env, str, &isCopy);
     strncpy (buff, copy_buff, nbuff);
+    buff[nbuff - 1] = '\0'; // ensure '\0' terminated
```
https://github.com/dotnet/runtime/issues/111923 added it to the TODO list there.

matouskozak commented on src/tasks/AndroidAppBuilder/Templates/monodroid.c:
```diff
@@ -304,20 +316,47 @@ mono_droid_runtime_init (const char* executable, int managed_argc, char* managed
+    {
+        mono_assembly_close (g_assembly);
+        g_assembly = NULL;
+    }
```
Do you know which API would do that job? I was looking for that in the Mono exports and only found 
https://github.com/dotnet/runtime/blob/256f32047c792224c0a018ca2a724cd5ec4cdf78/src/mono/mono/metadata/appdomain.c#L346-L349

https://github.com/dotnet/runtime/blob/59f2833b98e15e0831be8ad24fb321ee7bf9bd54/src/mono/mono/metadata/external-only.c#L326-L329

https://github.com/dotnet/runtime/blob/59f2833b98e15e0831be8ad24fb321ee7bf9bd54/src/mono/mono/metadata/external-only.c#L396-L400

and none of these looks right to me.

ivanpovazan replied:
Yeah, we don't seem to have a public API for that and I think it is the same principle as with `coreclr_shutdown` that it is not designed that you can reuse the domain, etc. 
I wouldn't worry about it that much, it was more of a question. Thanks!

matouskozak committed:
address review feedback by Aaron and Ivan

matouskozak commented:
/azp run runtime-android,runtime-androidemulator

ivanpovazan commented on src/tasks/AndroidAppBuilder/Templates/monodroid-librarymode.c:
```diff
@@ -15,7 +15,13 @@ void
 
 int
-Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_dir, jstring j_cache_dir, jstring j_testresults_dir, jstring j_entryPointLibName, jobjectArray j_args, long current_local_time);
+Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_dir, jstring j_cache_dir, jstring j_testresults_dir, jstring j_entryPointLibName, long current_local_time);
```
I think this is breaking the library mode tests.

```suggestion
Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_dir, jstring j_entryPointLibName, long current_local_time);
```

matouskozak replied:
Yes, forgot this one. Thank you.

matouskozak committed:
Update src/tasks/AndroidAppBuilder/Templates/monodroid-librarymode.c

Co-authored-by: Ivan Povazan <55002338+ivanpovazan@users.noreply.github.com>

matouskozak commented:
/azp run runtime-android,runtime-androidemulator
